src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java

```diff
@@ -260,26 +264,34 @@ boolean uploadsInProgress() {
     }
   }
 
+  private static String newResourceName(String instanceName, UUID uuid, HashCode hash, long size) {
+    String resourceName = format("uploads/%s/blobs/%s/%d", uuid, hash, size);
+    if (!Strings.isNullOrEmpty(instanceName)) {
+      resourceName = instanceName + "/" + resourceName;
+    }
+    return resourceName;
+  }
+
   /** Starts a file upload an returns a future representing the upload. */
-  private ListenableFuture<Void> startAsyncUpload(Chunker chunker) {
+  private ListenableFuture<Void> startAsyncUpload(HashCode hash, Chunker chunker) {
     try {
       chunker.reset();
     } catch (IOException e) {
       return Futures.immediateFailedFuture(e);
     }
 
-    SettableFuture<Void> currUpload = SettableFuture.create();
+    UUID uploadId = UUID.randomUUID();
```

Comment:
move this into newResourceName?

Reply:
I'd rather have newResourceName be deterministic and pure-functional here (will also change its name to just uploadResourceName) - the creation of the upload key should be governed by the caller.